torch/distributed/optim/zero_redundancy_optimizer.py

```diff
@@ -119,6 +128,78 @@ def main_hook(self):
         self.zero.step()
 
 
+class _DDPBucket(NamedTuple):
+    r"""
+    This contains the model parameters corresponding to a
+    :class:`DistributedDataParallel` gradient bucket.
+
+        bucket_index (int): index of the bucket determined by the DDP gradient
+            bucket all-reduce order.
+        params (List[torch.Tensor]): model parameters in the bucket.
+    """
+    bucket_index: int
+    params: List[torch.Tensor]
+
+
+class _OverlapStatus(enum.IntEnum):
+    r"""
+    This defines the three possible statuses that
+    :class:`ZeroRedundancyOptimizer` can be in when overlapping with
+    :class:`DistributedDataParallel`.
+
+        ``UNINITIALIZED``: The ZeRO instance is effectively uninitialized and
+            is waiting for DDP to finalize its bucketing.
+        ``DDP_HAS_REBUILT_BUCKETS``: DDP has rebuilt its buckets, meaning that
+            its bucketing is finalized. The ZeRO instance can now collect the
+            necessary information about the DDP bucketing.
+        ``INITIALIZED``: The ZeRO instance is fully initialized and can now
+            optimize parameters.
+    """
+    UNINITIALIZED = 0
+    DDP_HAS_REBUILT_BUCKETS = 1
+    INITIALIZED = 2
+
+
+class _OverlapInfo():
+    r"""
+    This contains the information needed by :class:`ZeroRedundancyOptimizer`
+    to overlap with :class:`DistributedDataParallel`.
+
+        status (_OverlapStatus): current status; see :class:`_OverlapStatus`
+            for more information.
+        params_per_bucket (List[List[torch.Tensor]]): ``params_per_bucket[i]``
+            gives the model parameters in the ``i``th bucket.
+        params_per_rank (List[List[torch.Tensor]]): ``params_per_rank[i]``
+            gives the model parameters assigned to the ``i``th rank, where the
+            parameters are grouped by increasing bucket indices.
+        offsets (Dict[int, int]): maps from bucket index to the offset in
+            ``self.params_per_rank[rank]`` giving the index of the first
+            parameter in that bucket, where ``rank`` is this process's own
+            rank; the keys of this :class:`dict` are the bucket indices
+            assigned to this rank.
+        broadcast_handles (List[Work]): :class:`list` of async work handles for
+            the parameter broadcasts.
+        bucket_index_to_future (Dict[int, torch.futures.Future]):
```

Comment:
nit (can be done in followup PRs): looks like each Future in this dict is just read once, so this cache probably won't save on Python-CPP context switch cost. Maybe this can be deleted, and directly use `bucket.get_future()`?

Reply:
In general, I would try to avoid the state vars unless that would give us either perf gain or considerable code simplification. As more vars might make it harder to understand and maintain.

Reply:
I may be misunderstanding, but the reason I stored this `dict` is to have a reference to all relevant `Future`s when running the hook for the last bucket to be able to call `wait()` on them. `bucket` is of type `GradBucket`, which does not offer a `get_future()` method. Rather, `c10d.ProcessGroup.work` objects provide a `get_future()` method. However, at the moment, this seems a bit backward. The `hook` passed into `hook_then_zero_step()` (most typically `allreduce_hook()`) returns a `Future`, where that `Future` is exactly gotten via `dist.all_reduce().get_future()`. Either way, I do not see how to get around storing at least something per bucket to update that gives access to the bucket's corresponding `Future` to give the rank access to the `Future` when executing the hook for the last bucket.